### PR TITLE
rails project

### PR DESCRIPTION
**Pull request by sehrnadeem1**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# Turbolinks makes following links in your web application faster. Read more: https://github.com/rails/turbolinks
+gem 'turbolinks'
+# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+gem 'jbuilder', '~> 2.0'
+# bundle exec rake doc:rails generates the API under doc/api.
+gem 'sdoc', '~> 0.4.0', group: :doc
+
+gem 'devise', '~> 3.4.0'
+
+gem 'pry'
```
Comment (sehrnadeem1) [inline Gemfile]: gem version

File: app/assets/javascripts/application.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//
+// It's not advisable to add code directly here, but if you do, it'll appear at the bottom of the
+// compiled file.
+//
+// Read Sprockets README (https://github.com/sstephenson/sprockets#sprockets-directives) for details
+// about supported directives.
+//
+//= require jquery
+//= require bootstrap.min.js
+//= require jquery_ujs
```
Comment (sehrnadeem1) [inline app/assets/javascripts/application.js]: verify

File: app/controllers/items_controller.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
+class ItemsController < ApplicationController
+  load_and_authorize_resource
+
+  def create                #POST /items/:id
```
Comment (HassanAkbar) [inline app/controllers/items_controller.rb]: #POST /items/:id
def create

File: app/controllers/items_controller.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    if @item.save
+      respond_to do |format|
+        format.html { redirect_to @item, notice: 'Item successfully added.' }
```
Comment (HassanAkbar) [inline app/controllers/items_controller.rb]: flash[:notice]

File: app/controllers/items_controller.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def destroy               #DELETE /items/:id
+    if @item.destroy
+      flash[:notice] = 'Item successfully deleted'
+    else
+      flash[:alert] = 'Item could not be deleted'
+    end
+    respond_to do |format|
+        format.html { redirect_to items_path }
```
Comment (HassanAkbar) [inline app/controllers/items_controller.rb]: space

File: app/controllers/order_items_controller.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
+class OrderItemsController < ApplicationController
+  load_and_authorize_resource :order
+  load_and_authorize_resource :order_item, through: :order
+
+  def new                   #GET    /order/:order_id/order_items/new
+    @items = Item.all
+    respond_to do |format|
+      format.js
+      format.html
```
Comment (HassanAkbar) [inline app/controllers/order_items_controller.rb]: remove

File: app/controllers/order_items_controller.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  def edit                  #GET    /order/:order_id/order_items/:id/edit
+    @items = Item.all
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def create                #POST   /order/:order_id/order_items
+    @items = Item.all
+    if @order_item.item.quantity >= @order_item.quantity
```
Comment (HassanAkbar) [inline app/controllers/order_items_controller.rb]: transaction

File: app/controllers/order_items_controller.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        flash[:alert] = "Order item could not be added beacuse of #{@order_item.errors.full_messages}."
+        respond_to do |format|
+          format.js
+          format.html { render 'new' }
+        end
+      end
+    else
+      flash[:alert] = 'Order item quantity cannot be greater than item stock quantity.'
+      respond_to do |format|
+          format.js
```
Comment (HassanAkbar) [inline app/controllers/order_items_controller.rb]: space

File: app/controllers/order_items_controller.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      else
+        flash.now[:alert] = 'Order item could not be updated.'
+        respond_to do |format|
+          format.html { render 'edit' }
+        end
+      end
+    else
+      flash[:alert] = 'Order items could not be updated, not enough items in stock.'
+      respond_to do |format|
+          format.html { render 'edit' }
```
Comment (HassanAkbar) [inline app/controllers/order_items_controller.rb]: space

File: app/controllers/order_items_controller.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def check_updated_quantity
+    if (params[:order_item][:quantity]).to_i > @order_item.quantity
+      quantity_diff  = params[:order_item][:quantity].to_i - @order_item.quantity
+      @order_item.item.update(quantity: @order_item.item.quantity - quantity_diff)
+    else
+      quantity_diff  = @order_item.quantity - params[:order_item][:quantity].to_i
+      @order_item.item.update(quantity: @order_item.item.quantity + quantity_diff)
+    end
```
Comment (HassanAkbar) [inline app/controllers/order_items_controller.rb]:
``` ruby 
order_item_quantity = @order_item.item.quantity - quantity_diff
@order_item.item.update(quantity: order_item_quantity)

File: app/controllers/orders_controller.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      end
+    end
+  end
+
+  def create                #POST   /user/:user_id/orders
+    if current_user.admin?
+      @order.user = current_user
+    end
+    if @order.save
+      #binding.pry
```
Comment (HassanAkbar) [inline app/controllers/orders_controller.rb]: look into admin problem.
remove

File: app/controllers/orders_controller.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def destroy               #DELETE /user/:user_id/orders/:id
+    if @order.destroy
+      flash[:notice] = "Order deleted."
+    else
+      flash[:alert] = "Order could not be deleted."
+    end
+    respond_to do |format|
+        format.html { redirect_to user_orders_path(@user) }
```
Comment (HassanAkbar) [inline app/controllers/orders_controller.rb]: space

File: app/controllers/waiters_controller.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+class WaitersController < ApplicationController
+  before_action :authenticate_user!
+
+  def index                   #GET    /waiters
+    if current_user.admin?
+      @waiters = User.waiter
+
+       respond_to do |format|
+        format.html
+        format.xml { render xml: @waiters }
```
Comment (HassanAkbar) [inline app/controllers/waiters_controller.rb]: remove

File: app/models/item.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+class Item < ActiveRecord::Base
+  validates :quantity, :numericality => { :greater_than_or_equal_to => 0 }
+  validates :price, :numericality => { :greater_than => 0 }
+  validates_presence_of :name
```
Comment (HassanAkbar) [inline app/models/item.rb]: validation

File: app/models/order.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
+class Order < ActiveRecord::Base
+  belongs_to :user
+  has_many :order_items, dependent: :destroy
+
+  validate :delivery_time_cannot_be_earlier_than_present
+
+  def delivery_time_cannot_be_earlier_than_present
+    if delivery_time.present?
+      input_time = delivery_time.change(:min => 0)
```
Comment (HassanAkbar) [inline app/models/order.rb]: min: 0

File: app/models/order.rb [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+      current_time = (DateTime.now).change(:min => 0)
```
Comment (HassanAkbar) [inline app/models/order.rb]:
``` ruby
current_time = DateTime.current.change(min: 0)

File: app/models/user.rb [hunk 17]
```diff
@@ -0,0 +1,10 @@
+class User < ActiveRecord::Base
+  enum role_id: {"admin"=>1, "waiter"=>2}
```
Comment (HassanAkbar) [inline app/models/user.rb]: enum role_id: {"admin" => 1, "waiter" => 2}

File: app/views/layouts/application.html.erb [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+<head>
+  <title>Restaurant Management System</title>
+
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
+
+  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
+
+
+  <%= stylesheet_link_tag    'application', media: 'all' %>
```
Comment (HassanAkbar) [inline app/views/layouts/application.html.erb]: space

File: app/views/orders/_form.html.erb [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    <% end %>
+    <div class="form-group">
+      <p>
+        <%= f.label :description %><br>
+        <%= f.text_field :description, class: 'form-control' %>
+      </p>
+
+      <p>
+        <%= f.label :status %><br>
+        <%= f.select :status, options_for_select([['Pending','Pending'],['In progress','In progress'],['Compeleted','Compeleted']]), {}, {class: 'form-control'} %>
```
Comment (HassanAkbar) [inline app/views/orders/_form.html.erb]: move to helper

**Pull request merged**